### ISSUE TITLE
A possible cause of duplicated case objects bug

### ISSUE THREAD

**Issue by arnor2000** [flink-extended/flink-scala-api#187]
Hi @novakov-alexey,
I think I have pinpointed a major cause of the `scala.MatchError: None (of class scala.None$)` bug, it actually touches all case objects by having several instances of what should be singleton. Other users have experienced this bug and may be related to #148, #106.

Another manifestation on a common Scala case object is on `Nil`:
```
java.util.NoSuchElementException: head of empty list
	at scala.collection.immutable.Nil$.head(List.scala:469)
	at scala.collection.immutable.Nil$.head(List.scala:466)
```

The problem appears only when we are removing `flink-scala_2.12-*.jar` from the Flink distribution (`lib/flink-scala_2.12-*.jar`) as Flink recommends it (ie: https://flink.apache.org/2022/02/22/scala-free-in-one-fifteen/). The reason is this jar contains a lot of custom Kryo serializers to handle Scala gracefully, notably [FlinkScalaKryoInstantiator](https://github.com/apache/flink/blob/91d81c427aa6312841ca868d54e8ce6ea721cd60/flink-scala/src/main/scala/org/apache/flink/runtime/types/FlinkScalaKryoInstantiator.scala#L44) which use Chill and allows Kryo to handle singleton `object`s and other Scala specificities.

Now, the question is: why Kryo continues to be involved in serialization despite the fact we shouldn't use it anymore thanks to flink-scala-api "no silent fallback to Kryo" feature?

Because of this constructor: `org.apache.flink.api.common.state.ValueStateDescriptor.ValueStateDescriptor(java.lang.String, java.lang.Class<T>)`, same for MapStateDescriptor, etc.
![Screenshot-ValueStateDescriptor](https://github.com/user-attachments/assets/cea28927-f268-469d-b8c5-29f0f59e5bbd)

It is absolutely not documented these constructors with a `Class` param is the door wide open to Kryo. And flink-scala-api can do nothing about it because it's not part of Scala API.

At least it should be documented in Migration section to switch from *StateDescription constructors with `Class` to `TypeInformation` param. I hope it can prevent weeks of headaches (in my case) to other users!

Thanks!

**Comment by arnaud-daroussin** [comment]
To give more details, I discovered the root cause of my duplicated case object problem by logging a stacktrace in the body of one of my case object:
```
java.lang.Exception: this exception is logged at TotalCount init
	at com.xxx.aggregations.AggregationFunction$TotalCount$.<init>(AggregationFunction.scala:96)
	at com.xxx.aggregations.AggregationFunction$TotalCount$ConstructorAccess.newInstance(Unknown Source)
	at com.esotericsoftware.kryo.Kryo$DefaultInstantiatorStrategy$1.newInstance(Kryo.java:1193)
	at com.esotericsoftware.kryo.Kryo.newInstance(Kryo.java:1061)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.createCopy(FieldSerializer.java:620)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.copy(FieldSerializer.java:624)
	at com.esotericsoftware.kryo.Kryo.copy(Kryo.java:862)
	at com.esotericsoftware.kryo.serializers.UnsafeCacheFields$UnsafeObjectField.copy(UnsafeCacheFields.java:297)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.copy(FieldSerializer.java:634)
	at com.esotericsoftware.kryo.Kryo.copy(Kryo.java:862)
	at com.esotericsoftware.kryo.serializers.UnsafeCacheFields$UnsafeObjectField.copy(UnsafeCacheFields.java:297)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.copy(FieldSerializer.java:634)
	at com.esotericsoftware.kryo.Kryo.copy(Kryo.java:862)
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.copy(DefaultArraySerializers.java:390)
	at com.esotericsoftware.kryo.serializers.DefaultArraySerializers$ObjectArraySerializer.copy(DefaultArraySerializers.java:289)
	at com.esotericsoftware.kryo.Kryo.copy(Kryo.java:862)
	at com.esotericsoftware.kryo.serializers.UnsafeCacheFields$UnsafeObjectField.copy(UnsafeCacheFields.java:297)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.copy(FieldSerializer.java:634)
	at com.esotericsoftware.kryo.Kryo.copy(Kryo.java:862)
	at com.esotericsoftware.kryo.serializers.UnsafeCacheFields$UnsafeObjectField.copy(UnsafeCacheFields.java:297)
	at com.esotericsoftware.kryo.serializers.FieldSerializer.copy(FieldSerializer.java:634)
	at com.esotericsoftware.kryo.Kryo.copy(Kryo.java:862)
	at org.apache.flink.api.java.typeutils.runtime.kryo.KryoSerializer.copy(KryoSerializer.java:298)
	at org.apache.flink.api.common.typeutils.base.MapSerializer.copy(MapSerializer.java:112)
	at org.apache.flink.runtime.state.HeapBroadcastState.<init>(HeapBroadcastState.java:70)
	at org.apache.flink.runtime.state.HeapBroadcastState.deepCopy(HeapBroadcastState.java:88)
	at org.apache.flink.runtime.state.HeapBroadcastState.deepCopy(HeapBroadcastState.java:41)
	at org.apache.flink.runtime.state.DefaultOperatorStateBackendSnapshotStrategy.syncPrepareResources(DefaultOperatorStateBackendSnapshotStrategy.java:96)
	at org.apache.flink.runtime.state.DefaultOperatorStateBackendSnapshotStrategy.syncPrepareResources(DefaultOperatorStateBackendSnapshotStrategy.java:40)
	at org.apache.flink.runtime.state.SnapshotStrategyRunner.snapshot(SnapshotStrategyRunner.java:77)
	at org.apache.flink.runtime.state.DefaultOperatorStateBackend.snapshot(DefaultOperatorStateBackend.java:230)
	at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.snapshotState(StreamOperatorStateHandler.java:242)
	at org.apache.flink.streaming.api.operators.StreamOperatorStateHandler.snapshotState(StreamOperatorStateHandler.java:185)
	at org.apache.flink.streaming.api.operators.AbstractStreamOperator.snapshotState(AbstractStreamOperator.java:348)
	at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.checkpointStreamOperator(RegularOperatorChain.java:228)
	at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.buildOperatorSnapshotFutures(RegularOperatorChain.java:213)
	at org.apache.flink.streaming.runtime.tasks.RegularOperatorChain.snapshotState(RegularOperatorChain.java:192)
	at org.apache.flink.streaming.runtime.tasks.SubtaskCheckpointCoordinatorImpl.takeSnapshotSync(SubtaskCheckpointCoordinatorImpl.java:720)
	at org.apache.flink.streaming.runtime.tasks.SubtaskCheckpointCoordinatorImpl.checkpointState(SubtaskCheckpointCoordinatorImpl.java:352)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.lambda$performCheckpoint$16(StreamTask.java:1369)
	at org.apache.flink.streaming.runtime.tasks.StreamTaskActionExecutor$1.runThrowing(StreamTaskActionExecutor.java:50)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.performCheckpoint(StreamTask.java:1357)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.triggerCheckpointOnBarrier(StreamTask.java:1314)
	at org.apache.flink.streaming.runtime.io.checkpointing.CheckpointBarrierHandler.notifyCheckpoint(CheckpointBarrierHandler.java:147)
	at org.apache.flink.streaming.runtime.io.checkpointing.CheckpointBarrierTracker.triggerCheckpointOnAligned(CheckpointBarrierTracker.java:301)
	at org.apache.flink.streaming.runtime.io.checkpointing.CheckpointBarrierTracker.processBarrier(CheckpointBarrierTracker.java:141)
	at org.apache.flink.streaming.runtime.io.checkpointing.CheckpointedInputGate.handleEvent(CheckpointedInputGate.java:181)
	at org.apache.flink.streaming.runtime.io.checkpointing.CheckpointedInputGate.pollNext(CheckpointedInputGate.java:159)
	at org.apache.flink.streaming.runtime.io.AbstractStreamTaskNetworkInput.emitNext(AbstractStreamTaskNetworkInput.java:122)
	at org.apache.flink.streaming.runtime.io.StreamOneInputProcessor.processInput(StreamOneInputProcessor.java:65)
	at org.apache.flink.streaming.runtime.io.StreamMultipleInputProcessor.processInput(StreamMultipleInputProcessor.java:85)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.processInput(StreamTask.java:579)
	at org.apache.flink.streaming.runtime.tasks.mailbox.MailboxProcessor.runMailboxLoop(MailboxProcessor.java:231)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.runMailboxLoop(StreamTask.java:909)
	at org.apache.flink.streaming.runtime.tasks.StreamTask.invoke(StreamTask.java:858)
	at org.apache.flink.runtime.taskmanager.Task.runWithSystemExitMonitoring(Task.java:958)
	at org.apache.flink.runtime.taskmanager.Task.restoreAndInvoke(Task.java:937)
	at org.apache.flink.runtime.taskmanager.Task.doRun(Task.java:751)
	at org.apache.flink.runtime.taskmanager.Task.run(Task.java:566)
	at java.base/java.lang.Thread.run(Thread.java:829)
```
I was surprised to see Kryo was involved in state serialization and was happily instantiating my case object when writing checkpoints.

It was due to usage of:
```scala
  val BroadcastStateDescriptor: MapStateDescriptor[Long, ApiMessage] =
    new MapStateDescriptor("received-apiMessages", classOf[Long], classOf[ApiMessage])
```
instead of:
```scala
  val BroadcastStateDescriptor: MapStateDescriptor[Long, ApiMessage] =
    new MapStateDescriptor("received-apiMessages",
      implicitly[TypeInformation[Long]],
      implicitly[TypeInformation[ApiMessage]])
```
For years, me and my colleagues didn't know Kryo was used for state serialization because of these constructors with `Class` params, this information should be widely shared!

Also note that we are using a `HashMapStateBackend`, this behavior may be different with other state backends.

**Comment by novakov-alexey** [comment]
Hi @arnor2000 

Thanks a lot for digging into it deeper. This is really good finding. I think this will also help other users of Scala API, indeed.

As I understood the confusion comes when `flink-scala_2.12-*.jar` is removed from a Flink job, but user code still uses **Java API constructors** which may lead to Kryo usage, like in case of `ValueStateDescriptor(...)`. 

Although, you have correctly highlighted that it promises to not fallback silently to Kryo, it can still happen even for Scala data classes when user code uses Java-oriented constructors. 

I see the following actions to improve the situation:

1. Try to prevent Kryo usage even from Java-oriented constructors. Not sure if it feasible at all. Perhaps some Java Reflection needs to be used already in runtime to report some meaningful error message.
2. Alternatively, update README to say that this library does not handle Kryo usage when user code uses Java-oriented constructors like the one you mentioned. 

What do you think?

**Comment by arnaud-daroussin** [comment]
Exactly!

In our case it was an unintended usage of Kryo, but I suppose duplication of case objects will also occur with explicit usage of Kryo with `TypeInformation.of(classOf[Option[String]]`, it can be really error-prone. It's worth to add a warning of Kryo usage in case of `flink-scala_2.12-*.jar` removal.

There is several borderline usage of Kryo in examples like in `FraudDetector`: `TypeInformation.of(classOf[Alert])` and `TypeInformation.of(classOf[KeyedFraudState])`. If these case classes were using an `Option` somewhere, any `None` value will be candidate for duplication.

About the actions you propose, the first one is the best but seems hard to achieve. I was thinking to extends `*StateDescriptor` with our own `*ScalaStateDescriptor` to hide constructors with `Class` params, but the only place to enforce it with current Scala API is `org.apache.flinkx.api.DataStream#broadcast(broadcastStateDescriptors: MapScalaStateDescriptor[_, _]*)`. Efficiently enforce it would require to provide our own `RuntimeContext` so it looks like to be a burden. Or maybe registering a special `TypeInfoFactory` on `Product` using `org.apache.flink.api.java.typeutils.TypeExtractor.registerFactory` as "Java-oriented" [StateDescriptor constructors](https://github.com/apache/flink/blob/84444d5071984a8543a81574dc69c0c9849a8a4a/flink-core/src/main/java/org/apache/flink/api/common/state/StateDescriptor.java#L161) are using it, but I don't know the implications... 

Clearly document the consequences of removing `flink-scala_2.12-*.jar` on Kryo could be a good first step for sure.

**Comment by novakov-alexey** [comment]
This library never claimed that it can prevent usage of Kryo if **user code does not use** it's serlializers via:
```scala
import org.apache.flinkx.api.serializers._
// and then
implicitly[TypeInformation[Long]]
```
Agree?

This is the example when a user does not use proper Scala serializers:
```scala
new MapStateDescriptor("received-apiMessages", classOf[Long], classOf[ApiMessage])
```
MapStateDescriptor is Java class, not Scala class.

So what you are saying is a user will be confused, because README says there is no silent fallback to Kryo, but actually there is. Since user opted-out from using proper serialisers, so now he is on its own. This is a user fault.

Of course we can say that this library could do better or its documentation could be better. So **first** we need to clarify this moment in docs.

**Secondly**, we need to describe the issue with "duplication of case objects" and why it happens.

**Thirdly**, we could try to improve library code to intercept case object serialization and then either fail-fast (3.1) or turn entire seralization process to use proper serializers (3.2). I am not sure if this 3rd point is doable including its options 3.1 and 3.2.

**Comment by arnaud-daroussin** [comment]
Hi @novakov-alexey,

I think the documentation work is done!

About the code improvement, the ideal would be to handle typing with `Class`.

One solution could be to register all types brought by flink-scala-api in `TypeExtractor.registerFactory()` (available only from Flink 1.19).

As for the TypeInformations generated for case classes, the call to `registerFactory()` could be made at the end of the code generated by the macro because all the necessary information is there. For example, here is the decompiled Java code of the constructor of object ApiMessage:

```java
TypeInformation apimessageTypeclass$macro$30 = var35.join(var38, var40, ((TypeTags)$u).TypeTag().apply((Mirror)$m, new ApiMessage$$typecreator29$1()));
this.info = apimessageTypeclass$macro$30;
this.bitmap$init$0 = (byte)(this.bitmap$init$0 | 32);
```
Although I have no idea how to do this with Magnolia.

Another possibility would be to use the cache (`org.apache.flinkx.api.LowPrioImplicits#cache`) filled at compilation by the generation of case classes, and generate with another macro the code that will do the registration at runtime. However, I have no idea how to ensure that this last macro is generated after all the macros that generated the case classes.

In short, I'm trying to propose ideas, but solution doesn't seem obvious!

**Comment by arnaud-daroussin** [comment]
Hi @novakov-alexey,
I don't think what I proposed in previous comment can be implemented in a reliable way.

So there is the fail-fast alternative which seems easier to implement:
- write a `FailFastTypeInfoFactory` throwing an `InvalidTypesException` with a very descriptive message indicating to use TypeInformation instead of Class.
- register this TypeInfoFactory on several common Scala classes: `Option`, `Either`, `Product`, `Array`, `List`, `Seq`, `Set`, `Map`.

How and when to register?
- with the "dirty" solution: register using `TypeExtractor.registerFactory()` static method. It's dirty because this method is tagged as `@Internal`. Because the method is static, it can be called as early as possible in object `serializers` for example.
- injecting the configuration: because we own `StreamExecutionEnvironment` and its `get*Environment()` and `create*Environment()` methods, we can inject the TypeInfoFactory registration into the configuration. Something like:
```scala
  def getExecutionEnvironment(configuration: Configuration): StreamExecutionEnvironment = {
    val serializationOption = ConfigOptions.key("pipeline.serialization-config").stringType().asList().noDefaultValue()
    val serializationConfig = configuration.getOptional(serializationOption).orElseGet(util.ArrayList[String])
    serializationConfig.add("scala.Option: {type: typeinfo, class: org.apache.flinkx.api.typeinfo.FailFastTypeInfoFactory}")
    configuration.set(serializationOption, serializationConfig)
  new StreamExecutionEnvironment(JavaEnv.getExecutionEnvironment(configuration))
}
```

Both solutions work from Flink 1.19. As far I know before Flink 1.19, the only way to register a TypeInfoFactory was with a `@TypeInfo` annotation on the type itself, so it's not an option.

A potential problem I see with the second solution is when `StreamExecutionEnvironment.getExecutionEnvironment()` is called: if it's later than a static declaration of a `*StateDescriptor` or a `TypeInformation.of`, the registration of our `FailFastTypeInfoFactory` will happen to late, am I right?

Of course, I think this registration should be an opt-out: enabled by default but can be disabled with a feature flag.

What do you think of this? Which option seems better?

If one of these solutions is good for you, I can propose a PR.

**Comment by arnaud-daroussin** [comment]
Yes, the documentation and the fail-fast feature address the issue, you can close it 👍

**Comment by arnor2000** [comment]
Yes, the documentation and the fail-fast feature address the issue, I'm closing it 👍

**Comment by arnor2000** [comment]
Hi,

I'm sorry to have to reopen the issue: my PR introducing fail-fast create 2 problems caused by the way I registered the `FailFastTypeInfoFactory`.

First, `MiniClusterWithClientResource` available in `flink-test-utils` dependency to do integration test with Flink registers 2 times the `FailFastTypeInfoFactory`, at `https://github.com/apache/flink/blob/release-1.20/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java#L93-L94` and 5 lines later at https://github.com/apache/flink/blob/release-1.20/flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/TestStreamEnvironment.java#L98 causing at test startup:
```
java.lang.IllegalArgumentException: Could not configure serializers from [scala.Product: {type: typeinfo, class: org.apache.flinkx.api.typeinfo.FailFastTypeInfoFactory}, scala.Option: {type: typeinfo, class: org.apache.flinkx.api.typeinfo.FailFastTypeInfoFactory}, scala.util.Either: {type: typeinfo, class: org.apache.flinkx.api.typeinfo.FailFastTypeInfoFactory}, scala.Array: {type: typeinfo, class: org.apache.flinkx.api.typeinfo.FailFastTypeInfoFactory}, scala.collection.Iterable: {type: typeinfo, class: org.apache.flinkx.api.typeinfo.FailFastTypeInfoFactory}].
	at org.apache.flink.api.common.serialization.SerializerConfigImpl.parseSerializationConfigWithExceptionHandling(SerializerConfigImpl.java:446)
	at org.apache.flink.api.common.serialization.SerializerConfigImpl.lambda$configure$6(SerializerConfigImpl.java:373)
	at java.base/java.util.Optional.ifPresent(Optional.java:178)
	at org.apache.flink.api.common.serialization.SerializerConfigImpl.configure(SerializerConfigImpl.java:373)
	at org.apache.flink.api.common.ExecutionConfig.configure(ExecutionConfig.java:1293)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.configure(StreamExecutionEnvironment.java:1104)
	at org.apache.flink.streaming.util.TestStreamEnvironment.lambda$setAsContext$0(TestStreamEnvironment.java:99)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.lambda$getExecutionEnvironment$9(StreamExecutionEnvironment.java:2645)
	at java.base/java.util.Optional.map(Optional.java:260)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.getExecutionEnvironment(StreamExecutionEnvironment.java:2645)
	at org.apache.flinkx.api.StreamExecutionEnvironment$.getExecutionEnvironment(StreamExecutionEnvironment.scala:831)
	at com.xxx.integrationTestUtils.flink.FlinkTestEnv.startEnv(FlinkTestEnv.scala:51)
	at com.xxx.integrationTestUtils.main.TestEnv.startEnv(TestEnv.scala:29)
	at com.xxx.integrationTestUtils.MainTestBase.beforeAll(MainTestBase.scala:30)
	at org.scalatest.BeforeAndAfterAll.liftedTree1$1(BeforeAndAfterAll.scala:212)
	at org.scalatest.BeforeAndAfterAll.run(BeforeAndAfterAll.scala:210)
	at org.scalatest.BeforeAndAfterAll.run$(BeforeAndAfterAll.scala:208)
	at com.xxx.integrationTestUtils.MainTestBase.run(MainTestBase.scala:20)
	at org.scalatest.tools.SuiteRunner.run(SuiteRunner.scala:47)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13(Runner.scala:1321)
	at org.scalatest.tools.Runner$.$anonfun$doRunRunRunDaDoRunRun$13$adapted(Runner.scala:1315)
	at scala.collection.immutable.List.foreach(List.scala:334)
	at org.scalatest.tools.Runner$.doRunRunRunDaDoRunRun(Runner.scala:1315)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24(Runner.scala:992)
	at org.scalatest.tools.Runner$.$anonfun$runOptionallyWithPassFailReporter$24$adapted(Runner.scala:970)
	at org.scalatest.tools.Runner$.withClassLoaderAndDispatchReporter(Runner.scala:1481)
	at org.scalatest.tools.Runner$.runOptionallyWithPassFailReporter(Runner.scala:970)
	at org.scalatest.tools.Runner$.run(Runner.scala:798)
	at org.scalatest.tools.Runner.run(Runner.scala)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.runScalaTest2or3(ScalaTestRunner.java:43)
	at org.jetbrains.plugins.scala.testingSupport.scalaTest.ScalaTestRunner.main(ScalaTestRunner.java:26)
Caused by: org.apache.flink.api.common.functions.InvalidTypesException: A TypeInfoFactory for type 'interface scala.Product' is already registered.
	at org.apache.flink.api.java.typeutils.TypeExtractor.registerFactory(TypeExtractor.java:164)
	at org.apache.flink.api.common.serialization.SerializerConfigImpl.parseAndRegisterTypeFactory(SerializerConfigImpl.java:532)
	at org.apache.flink.api.common.serialization.SerializerConfigImpl.parseSerializationConfig(SerializerConfigImpl.java:486)
	at org.apache.flink.api.common.serialization.SerializerConfigImpl.parseSerializationConfigWithExceptionHandling(SerializerConfigImpl.java:443)
	... 30 more
```

Second problem, `pipeline.serialization-config` used to register `FailFastTypeInfoFactory` is only available with the new default YAML config file introduced in [Flink 1.19](https://flink.apache.org/2024/03/18/announcing-the-release-of-apache-flink-1.19/#standard-yaml-for-flink-configuration), so if users still use old `flink-conf.yaml` and users in Flink 1.18, conf we injected in `pipeline.serialization-config` will be rejected:
```
org.apache.flink.client.program.ProgramInvocationException: The main method caused an error: pipeline.serialization-config is only supported with the standard YAML config parser, please use "config.yaml" as the config file.
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:372)
	at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222)
	at org.apache.flink.client.ClientUtils.executeProgram(ClientUtils.java:108)
	at org.apache.flink.client.cli.CliFrontend.executeProgram(CliFrontend.java:1026)
	at org.apache.flink.client.cli.CliFrontend.run(CliFrontend.java:247)
	at org.apache.flink.client.cli.CliFrontend.parseAndRun(CliFrontend.java:1270)
	at org.apache.flink.client.cli.CliFrontend.lambda$mainInternal$10(CliFrontend.java:1367)
	at org.apache.flink.runtime.security.contexts.NoOpSecurityContext.runSecured(NoOpSecurityContext.java:28)
	at org.apache.flink.client.cli.CliFrontend.mainInternal(CliFrontend.java:1367)
	at org.apache.flink.client.cli.CliFrontend.main(CliFrontend.java:1335)
Caused by: java.lang.UnsupportedOperationException: pipeline.serialization-config is only supported with the standard YAML config parser, please use "config.yaml" as the config file.
	at org.apache.flink.api.common.serialization.SerializerConfigImpl.configure(SerializerConfigImpl.java:377)
	at org.apache.flink.api.common.ExecutionConfig.configure(ExecutionConfig.java:1293)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.configure(StreamExecutionEnvironment.java:1104)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.<init>(StreamExecutionEnvironment.java:300)
	at org.apache.flink.client.program.StreamContextEnvironment.<init>(StreamContextEnvironment.java:106)
	at org.apache.flink.client.program.StreamContextEnvironment.lambda$setAsContext$4(StreamContextEnvironment.java:222)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.lambda$getExecutionEnvironment$9(StreamExecutionEnvironment.java:2645)
	at java.base/java.util.Optional.map(Optional.java:260)
	at org.apache.flink.streaming.api.environment.StreamExecutionEnvironment.getExecutionEnvironment(StreamExecutionEnvironment.java:2645)
	at org.apache.flinkx.api.StreamExecutionEnvironment$.getExecutionEnvironment(StreamExecutionEnvironment.scala:831)
	at com.xxx.main.Main$.loadConf$default$2(Main.scala:23)
	at com.xxx.main.Main$.main(Main.scala:17)
	at com.xxx.main.Main.main(Main.scala)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:355)
	... 9 more
```

We can work around these problems by opting out fail-fast with `DISABLE_FAIL_FAST_ON_SCALA_TYPE_RESOLUTION_WITH_CLASS` env var.

To fix these problems, we can register the fail-fast factory with the "dirty" solution explained [here](https://github.com/flink-extended/flink-scala-api/issues/187#issuecomment-2614119363): directly using `TypeExtractor.registerFactory()` static method. I think there will be no drawback, and we will have the full control over how and when the `TypeExtractor.registerFactory()` is called, which is not the case using current configuration to register them.

I can quickly propose a fix to switch to use `TypeExtractor.registerFactory()` static method, or we can roll back this PR.

**Comment by novakov-alexey** [comment]
Hi @arnor2000 , please create a fix using `TypeExtractor.registerFactory()` static method.

**Comment by arnaud-daroussin** [comment]
Hi @novakov-alexey,
I tested extensively a snapshot version and I don't see any drawback registering using static method.
I think we should supersede as soon as possible version 1.2.3 because the 2 problems I listed above will cause confusion if used.
What do you think to make a release for this?
Thanks

**Comment by novakov-alexey** [comment]
Hi @arnaud-daroussin ,

It is ready: https://github.com/flink-extended/flink-scala-api/releases/tag/v1.20.0_1.2.4
